File: libmuscle/python/libmuscle/test/test_outbox.py
from libmuscle.outbox import Outbox
from libmuscle.mcp.message import Message

from copy import copy
import pytest

from ymmsl import Reference


@pytest.fixture
def outbox():
    return Outbox()


@pytest.fixture
def message():
    Ref = Reference
    return Message(
            Ref('sender.out'), Ref('receiver.in'),
            None, 0.0, 1.0,
            bytes(),
            'testing'.encode('utf-8'))


def test_create_outbox():
    box = Outbox()
    assert box._Outbox__queue.qsize() == 0


def test_deposit_message(outbox, message):
    outbox.deposit(message)
    assert outbox._Outbox__queue.qsize() == 1
    assert outbox._Outbox__queue.get(message)


def test_retrieve_message(outbox, message):
    outbox._Outbox__queue.put(message)
    assert outbox.retrieve() == message


def test_deposit_retrieve_order(outbox, message):
    m1 = copy(message)
    m2 = copy(message)

    outbox.deposit(m1)
    outbox.deposit(m2)

    assert outbox.retrieve() == m1
    assert outbox.retrieve() == m2

File: libmuscle/cpp/src/libmuscle/tests/test_outbox.cpp
#include "libmuscle/outbox.hpp"

#include <memory>

#include <gtest/gtest.h>

#include <libmuscle/util.hpp>
#include <libmuscle/data.hpp>
#include <libmuscle/mpp_message.hpp>
#include <ymmsl/ymmsl.hpp>


using libmuscle::_MUSCLE_IMPL_NS::DataConstRef;
using libmuscle::_MUSCLE_IMPL_NS::Optional;
using libmuscle::_MUSCLE_IMPL_NS::Outbox;
using libmuscle::_MUSCLE_IMPL_NS::MPPMessage;
using ymmsl::Reference;


int main(int argc, char *argv[]) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

TEST(libmuscle_outbox, test_create_outbox) {
    Outbox box;

    ASSERT_TRUE(box.is_empty());
}

TEST(libmuscle_outbox, test_deposit_retrieve_message) {
    Outbox box;

    MPPMessage message(
            Reference("sender.out"), Reference("receiver.in"),
            Optional<int>(),
            0.0, 1.0,
            DataConstRef(),
            0, 1.0,
            DataConstRef("testing"));

    auto message_data = std::make_unique<DataConstRef>(message.encoded());

    auto msg_ptr = message_data.get();

    box.deposit(std::move(message_data));
    ASSERT_FALSE(box.is_empty());

    auto message2 = box.retrieve();
    ASSERT_TRUE(box.is_empty());
    ASSERT_EQ(message2.get(), msg_ptr);
}

